# Data Cleaning and Exploratory Data Analysis

Setting the stage

Load the datasets. 

In [143]:
pre <- "/Users/patrickokwir/Desktop/Lighthouse-data-notes/Jack-Talk-Surveys/datasets/CASE STUDY - Livestream pre FY23.csv"
post <- "/Users/patrickokwir/Desktop/Lighthouse-data-notes/Jack-Talk-Surveys/datasets/CASE STUDY - Livestream post FY23.csv"

In [180]:
# load dfs from csv file
library(readr)

df_pre <- read.csv(pre, header = TRUE, sep = ",")
df_post <- read.csv(post, header = TRUE, sep = ",")

### The pre surveys cleaning 

In [181]:
#head df_pre
head(df_pre)

Finished RecordedDate     Q1_a                       Q2_a             
1 TRUE     2022-09-21 13:13                                             
2 TRUE     2022-09-21 13:14 Strongly disagree          Agree            
3 TRUE     2022-09-21 13:14 Agree                      Somewhat agree   
4 TRUE     2022-09-21 13:15 Strongly disagree          Strongly disagree
5 TRUE     2022-09-21 13:15 Somewhat disagree          Disagree         
6 TRUE     2022-09-21 13:16 Neither agree nor disagree Agree            
  Q3_a                       Q4_a                       id      
1                                                               
2 Somewhat agree             Agree                      O_A_09_L
3 Neither agree nor disagree Neither agree nor disagree A_N_02_8
4 Somewhat agree             Neither agree nor disagree M_U_07_1
5 Agree                      Somewhat agree             A_E_06_A
6 Agree                      Agree                      Y_O_04_1

In [182]:
# return rows where finished = False
head(df_pre[df_pre$Finished == FALSE, ])

Finished RecordedDate     Q1_a Q2_a Q3_a Q4_a id      
17 FALSE    2022-09-22 13:11                             
18 FALSE    2022-09-22 13:11                             
19 FALSE    2022-09-22 13:12                             
20 FALSE    2022-09-22 13:12                     L_T_12_A
21 FALSE    2022-09-22 13:12                             
22 FALSE    2022-09-22 13:12

>>1. Remove incomplete responses from df_pre. These are rows where column 'Finished' = False.

In [183]:
# Remove incomplete responses from df_pre. These are rows where column 'Finished' = False.

# remove rows from df_pre where column Finished = False
# count rows where column Finished = False
count_false_finished <- sum(df_pre$Finished == FALSE)
count_false_finished
cat("There are:", count_false_finished, "incomplete responses in df_pre. These will be removed from the dataframe. \n")

[1] 2878

There are: 2878 incomplete responses in df_pre. These will be removed from the dataframe. 


`There are 2,878 responses not completed.`

In [184]:
# Remove rows where the user did not finish the survey
df_pre <- df_pre[df_pre$Finished != FALSE, ]
count_false_finished <- nrow(df_pre[df_pre$Finished == FALSE, ])
count_true_finished <- nrow(df_pre[df_pre$Finished != FALSE, ])

count_false_finished
cat(count_true_finished, "responses are marked as finished.")

[1] 0

8047 responses are marked as finished.

In [185]:
head(df_pre)

Finished RecordedDate     Q1_a                       Q2_a             
1 TRUE     2022-09-21 13:13                                             
2 TRUE     2022-09-21 13:14 Strongly disagree          Agree            
3 TRUE     2022-09-21 13:14 Agree                      Somewhat agree   
4 TRUE     2022-09-21 13:15 Strongly disagree          Strongly disagree
5 TRUE     2022-09-21 13:15 Somewhat disagree          Disagree         
6 TRUE     2022-09-21 13:16 Neither agree nor disagree Agree            
  Q3_a                       Q4_a                       id      
1                                                               
2 Somewhat agree             Agree                      O_A_09_L
3 Neither agree nor disagree Neither agree nor disagree A_N_02_8
4 Somewhat agree             Neither agree nor disagree M_U_07_1
5 Agree                      Somewhat agree             A_E_06_A
6 Agree                      Agree                      Y_O_04_1

`There are 8047 datapoints indicated as complete, there are still datapoints indicated as 'finished' that are not complete.
we have to remove the incomplete datapoints.`

In [191]:
library(dplyr)

# Filter rows where at least one of the specified columns is blank
blank_rows <- df_pre %>%
  filter(Q1_a == "" |
         Q2_a == "" |
         Q3_a == "" |
         Q4_a == "" |
         id == "")

# Filter rows where all of the specified columns are not blank
head(blank_rows)
#count rows
count_blank_rows <- nrow(blank_rows)
cat("There are: ", count_blank_rows, " all blank rows in the dataset")

Finished RecordedDate     Q1_a Q2_a Q3_a Q4_a id
1 TRUE     2022-09-21 13:13                       
2 TRUE     2022-10-03 9:08                        
3 TRUE     2022-10-03 9:09                        
4 TRUE     2022-10-03 9:11                        
5 TRUE     2022-10-03 9:14                        
6 TRUE     2022-10-03 9:14

There are:  582  all blank rows in the dataset

`There are 7574 responses in the dataset that would be considered complete.`

```python

>> 2. remove duplicate responses, keep only most recent response for each survey

In [192]:
# set df_pre to subset with non-blank rows
df_pre <- df_pre %>%
  filter(!(Q1_a == "" |
         Q2_a == "" |
         Q3_a == "" |
         Q4_a == "" |
         id == ""))

head(df_pre)

Finished RecordedDate     Q1_a                       Q2_a             
1 TRUE     2022-09-21 13:14 Strongly disagree          Agree            
2 TRUE     2022-09-21 13:14 Agree                      Somewhat agree   
3 TRUE     2022-09-21 13:15 Strongly disagree          Strongly disagree
4 TRUE     2022-09-21 13:15 Somewhat disagree          Disagree         
5 TRUE     2022-09-21 13:16 Neither agree nor disagree Agree            
6 TRUE     2022-09-21 13:16 Agree                      Agree            
  Q3_a                       Q4_a                       id      
1 Somewhat agree             Agree                      O_A_09_L
2 Neither agree nor disagree Neither agree nor disagree A_N_02_8
3 Somewhat agree             Neither agree nor disagree M_U_07_1
4 Agree                      Somewhat agree             A_E_06_A
5 Agree                      Agree                      Y_O_04_1
6 Somewhat agree             Somewhat agree             A_Y_06_A

In [198]:
# QC check
# check if the data contains any missing values or empty strings in the subset of columns. 
qc_check <- df_pre %>%
  filter(Q1_a == "" |
         Q2_a == "" |
         Q3_a == "" |
         Q4_a == "" |
         id == "")

head(qc_check)

Finished RecordedDate Q1_a Q2_a Q3_a Q4_a id

In [199]:
library(dplyr)

# Remove duplicates and keep most recent response for each data point
unique_recent_responses <- df_pre %>%
  arrange(id, desc(RecordedDate)) %>%
  distinct(id, .keep_all = TRUE)


In [200]:
head(unique_recent_responses)
cat ("There are ", nrow(unique_recent_responses), "unduplicated records in the dataset")

Finished RecordedDate     Q1_a                      
1 TRUE     2022-11-21 17:38 Agree                     
2 TRUE     2023-05-01 12:23 Strongly disagree         
3 TRUE     2023-05-30 8:49  Disagree                  
4 TRUE     2023-01-25 9:08  Strongly disagree         
5 TRUE     2023-05-15 14:05 Agree                     
6 TRUE     2023-05-18 9:31  Neither agree nor disagree
  Q2_a                       Q3_a              Q4_a              id      
1 Neither agree nor disagree Somewhat agree    Somewhat agree    ._O_04_A
2 Strongly disagree          Strongly disagree Strongly disagree 6_9_02_L
3 Disagree                   Disagree          Disagree          6_9_12_L
4 Strongly disagree          Strongly disagree Strongly disagree :_!_09_K
5 Agree                      Agree             Agree             A_A_01_A
6 Disagree                   Agree             Strongly agree    A_A_01_B

There are  6358 unduplicated records in the dataset

In [201]:
# Export df_pre to a CSV file
write.csv(df_pre, "/Users/patrickokwir/Desktop/Lighthouse-data-notes/Jack-Talk-Surveys/data_output/df_pre.csv", row.names = FALSE)

### The post surveys cleaning 

In [202]:
head(df_post)

Finished RecordedDate     Q1_b                       Q2_b          
1  TRUE    2022-09-21 13:55 Neither agree nor disagree Somewhat agree
2  TRUE    2022-09-21 13:57 Somewhat agree             Somewhat agree
3  TRUE    2022-09-21 13:57 Agree                      Disagree      
4  TRUE    2022-09-21 13:58 Agree                      Disagree      
5 FALSE    2022-09-24 13:14                                          
6 FALSE    2022-09-24 13:16                                          
  Q3_b                       Q4_b                      
1 Agree                      Somewhat agree            
2 Neither agree nor disagree Neither agree nor disagree
3 Somewhat agree             Agree                     
4 Strongly agree             Strongly agree            
5                                                      
6                                                      
  Q5                         Q6                        
1 Somewhat agree             Somewhat agree            
2 Agree                      Agree                     
3 Agree                      Agree                     
4 Neither agree nor disagree Neither agree nor disagree
5                                                      
6                                                      
  Q7                         Q8 id      
1 Neither agree nor disagree  5 A_E_06_A
2                             5 I_K_10_1
3 Neither agree nor disagree  7 A_N_09_D
4 Neither agree nor disagree  5 L_T_12_A
5                            NA         
6                            NA

>>1. Remove incomplete responses from df_post. These are rows where column 'Finished' = False.

In [204]:
# return rows where finished = False
head(df_post[df_post$Finished == FALSE, ])
nrow(df_post[df_post$Finished == FALSE, ])

#print the number of incomplete responses
cat("There are", nrow(df_post[df_post$Finished == FALSE, ]), "incomplete responses in the dataset\n")

Finished RecordedDate     Q1_b Q2_b Q3_b Q4_b Q5 Q6 Q7 Q8 id      
5  FALSE    2022-09-24 13:14                              NA         
6  FALSE    2022-09-24 13:16                              NA         
7  FALSE    2022-09-24 13:17                              NA         
8  FALSE    2022-09-24 13:20                              NA         
9  FALSE    2022-09-24 13:23                              NA         
10 FALSE    2022-09-24 13:31                              NA O_A_05_O

[1] 3040

There are 3040 incomplete responses in the dataset


In [205]:
df_post <- df_post[df_post$Finished == TRUE, ]
head(df_post)
# print number of complete records in the dataset
cat("There are", nrow(df_post), "completed records in the dataset.")

Finished RecordedDate     Q1_b                       Q2_b          
1  TRUE     2022-09-21 13:55 Neither agree nor disagree Somewhat agree
2  TRUE     2022-09-21 13:57 Somewhat agree             Somewhat agree
3  TRUE     2022-09-21 13:57 Agree                      Disagree      
4  TRUE     2022-09-21 13:58 Agree                      Disagree      
14 TRUE     2022-10-03 9:28  Strongly agree             Strongly agree
15 TRUE     2022-10-03 9:29  Strongly disagree          Strongly agree
   Q3_b                       Q4_b                      
1  Agree                      Somewhat agree            
2  Neither agree nor disagree Neither agree nor disagree
3  Somewhat agree             Agree                     
4  Strongly agree             Strongly agree            
14 Somewhat agree             Somewhat agree            
15 Somewhat agree             Somewhat disagree         
   Q5                         Q6                        
1  Somewhat agree             Somewhat agree            
2  Agree                      Agree                     
3  Agree                      Agree                     
4  Neither agree nor disagree Neither agree nor disagree
14 Agree                      Agree                     
15 Disagree                   Agree                     
   Q7                         Q8 id      
1  Neither agree nor disagree  5 A_E_06_A
2                              5 I_K_10_1
3  Neither agree nor disagree  7 A_N_09_D
4  Neither agree nor disagree  5 L_T_12_A
14 Agree                      10 Y_G_08_W
15 Agree                       2 A_N_11_D

There are 3816 completed records in the dataset.

In [207]:
# Check if more incomplete records are present in the dataset using subset of columns Q1_b,	Q2_b,	Q3_b, Q4_b,	Q5,	Q6,	Q7,	Q8

library(dplyr)

# Filter rows where all of the specified columns are blank
all_blank_subset_rows <- df_post %>%
  filter(Q1_b == "" |
         Q2_b == "" |
         Q3_b == "" |
         Q4_b == "" |
         Q5 == "" |
         Q6 == "" |
         Q7 == "" |
         is.na(Q8))

# show the first 10 rows
head(all_blank_subset_rows)

# print number of records 
cat("There are", nrow(all_blank_subset_rows), "records where all of the subset columns are blank")


Finished RecordedDate     Q1_b           Q2_b             
1 TRUE     2022-09-21 13:57 Somewhat agree Somewhat agree   
2 TRUE     2022-10-03 10:12 Somewhat agree Somewhat disagree
3 TRUE     2022-10-14 22:27                                 
4 TRUE     2022-10-14 22:29 Strongly agree Strongly agree   
5 TRUE     2022-10-28 17:54 Agree          Agree            
6 TRUE     2022-11-03 10:27                                 
  Q3_b                       Q4_b                      
1 Neither agree nor disagree Neither agree nor disagree
2 Strongly agree             Somewhat disagree         
3                                                      
4                            Strongly agree            
5 Agree                      Agree                     
6                                                      
  Q5                         Q6                Q7             Q8 id      
1 Agree                      Agree                             5 I_K_10_1
2 Strongly disagree          Strongly disagree                 2 E_I_04_O
3                                                             NA         
4 Neither agree nor disagree Strongly agree    Strongly agree  5 E_E_09_T
5                                                             NA O_D_03_H
6                                                             NA

There are 298 records where all of the subset columns are blank

In [208]:
# remove those from df_post

df_post <- df_post %>%
  filter(!(Q1_b == "" |
         Q2_b == "" |
         Q3_b == "" |
         Q4_b == "" |
         Q5 == "" |
         Q6 == "" |
         Q7 == "" |
         is.na(Q8)))

# show the first 10 rows
cat("There are ", nrow(df_post), " rows in df_post. \n")

There are  3518  rows in df_post. 


>> 2. remove duplicate responses, keep only most recent response for each survey

In [209]:
library(dplyr)

# Remove duplicates and keep most recent response for each data point
unique_recent_responses <- df_post %>%
  arrange(id, desc(RecordedDate)) %>%
  distinct(id, .keep_all = TRUE)

head(unique_recent_responses)
cat("Number of unique responses:", nrow(unique_recent_responses))

Finished RecordedDate     Q1_b              Q2_b                      
1 TRUE     2023-05-02 13:54 Strongly disagree Strongly disagree         
2 TRUE     2023-05-09 9:49  Strongly disagree Disagree                  
3 TRUE     2023-03-07 16:20 Strongly disagree Agree                     
4 TRUE     2023-05-01 13:02 Somewhat disagree Neither agree nor disagree
5 TRUE     2023-05-02 14:58 Agree             Somewhat agree            
6 TRUE     2023-04-24 17:01 Somewhat agree    Somewhat agree            
  Q3_b                       Q4_b                       Q5               
1 Strongly disagree          Neither agree nor disagree Strongly agree   
2 Strongly disagree          Strongly disagree          Strongly disagree
3 Somewhat agree             Somewhat disagree          Somewhat disagree
4 Neither agree nor disagree Neither agree nor disagree Agree            
5 Neither agree nor disagree Neither agree nor disagree Somewhat agree   
6 Neither agree nor disagree Neither agree nor disagree Somewhat agree   
  Q6                         Q7                Q8 id      
1 Strongly agree             Agree             10 6_E_07_A
2 Strongly disagree          Strongly disagree  0 A_8_12_1
3 Somewhat agree             Agree              9 A_A_01_A
4 Strongly agree             Strongly agree    10 A_A_01_C
5 Neither agree nor disagree Somewhat agree     7 A_A_01_D
6 Neither agree nor disagree Somewhat agree     7 A_A_01_M

Number of unique responses: 3226

In [210]:
#QC check
# check if there are still subset of columns with missing values
qc_check <- df_post %>%
  filter(Q1_b == "" |
         Q2_b == "" |
         Q3_b == "" |
         Q4_b == "" |
         Q5 == "" |
         Q6 == "" |
         Q7 == "" |
         is.na(Q8))

# show the first 10 rows
cat("There are ", nrow(qc_check), "rows with missing values. \n")

There are  0 rows with missing values. 


In [211]:
# Remove duplicates and keep most recent response for each survey
unique_recent_responses <- df_post %>%
  group_by(id) %>%
  arrange(desc(RecordedDate)) %>%
  slice(1) %>%
  ungroup()

cat("Number of unique responses:", nrow(unique_recent_responses), "\n")

Number of unique responses: 3226 


In [212]:
df_post <- unique_recent_responses
head(df_post)
cat("There are:", nrow(df_post), "unique complete responses in the dataset.")

Finished RecordedDate     Q1_b              Q2_b                      
1 TRUE     2023-05-02 13:54 Strongly disagree Strongly disagree         
2 TRUE     2023-05-09 9:49  Strongly disagree Disagree                  
3 TRUE     2023-03-07 16:20 Strongly disagree Agree                     
4 TRUE     2023-05-01 13:02 Somewhat disagree Neither agree nor disagree
5 TRUE     2023-05-02 14:58 Agree             Somewhat agree            
6 TRUE     2023-04-24 17:01 Somewhat agree    Somewhat agree            
  Q3_b                       Q4_b                       Q5               
1 Strongly disagree          Neither agree nor disagree Strongly agree   
2 Strongly disagree          Strongly disagree          Strongly disagree
3 Somewhat agree             Somewhat disagree          Somewhat disagree
4 Neither agree nor disagree Neither agree nor disagree Agree            
5 Neither agree nor disagree Neither agree nor disagree Somewhat agree   
6 Neither agree nor disagree Neither agree nor disagree Somewhat agree   
  Q6                         Q7                Q8 id      
1 Strongly agree             Agree             10 6_E_07_A
2 Strongly disagree          Strongly disagree  0 A_8_12_1
3 Somewhat agree             Agree              9 A_A_01_A
4 Strongly agree             Strongly agree    10 A_A_01_C
5 Neither agree nor disagree Somewhat agree     7 A_A_01_D
6 Neither agree nor disagree Somewhat agree     7 A_A_01_M

There are: 3226 unique complete responses in the dataset.

In [213]:
# Export df_post to csv 
write.csv(df_post, "/Users/patrickokwir/Desktop/Lighthouse-data-notes/Jack-Talk-Surveys/data_output/df_post.csv", row.names = FALSE)

>> ### 4. match any pre and post data from the same respondent using the provided IDs

In [215]:
#4.1 create a new column called 'Pre' on the df_pre dataframe and set to 'pre'
df_pre$Pre <- 'pre'

#4.2 create a new column called 'Post' on the df_post dataframe and set to 'post'
df_post$Post <- 'post'

In [218]:
# merge two dataframes into one on the column 'id'
merged_df <- merge(df_pre, df_post, by = 'id', all = TRUE)
head(merged_df)

id       Finished.x RecordedDate.x   Q1_a             
1 ._O_04_A TRUE       2022-11-21 17:38 Agree            
2 6_9_02_L TRUE       2023-05-01 12:23 Strongly disagree
3 6_9_12_L TRUE       2023-05-30 8:49  Disagree         
4 6_E_07_A   NA       NA               NA               
5 :_!_09_K TRUE       2023-01-25 9:08  Strongly disagree
6 A_8_12_1   NA       NA               NA               
  Q2_a                       Q3_a              Q4_a              Pre Finished.y
1 Neither agree nor disagree Somewhat agree    Somewhat agree    pre   NA      
2 Strongly disagree          Strongly disagree Strongly disagree pre   NA      
3 Disagree                   Disagree          Disagree          pre   NA      
4 NA                         NA                NA                NA  TRUE      
5 Strongly disagree          Strongly disagree Strongly disagree pre   NA      
6 NA                         NA                NA                NA  TRUE      
  RecordedDate.y   Q1_b              Q2_b              Q3_b             
1 NA               NA                NA                NA               
2 NA               NA                NA                NA               
3 NA               NA                NA                NA               
4 2023-05-02 13:54 Strongly disagree Strongly disagree Strongly disagree
5 NA               NA                NA                NA               
6 2023-05-09 9:49  Strongly disagree Disagree          Strongly disagree
  Q4_b                       Q5                Q6               
1 NA                         NA                NA               
2 NA                         NA                NA               
3 NA                         NA                NA               
4 Neither agree nor disagree Strongly agree    Strongly agree   
5 NA                         NA                NA               
6 Strongly disagree          Strongly disagree Strongly disagree
  Q7                Q8 Post
1 NA                NA NA  
2 NA                NA NA  
3 NA                NA NA  
4 Agree             10 post
5 NA                NA NA  
6 Strongly disagree  0 post

In [219]:
# create a new column called 'Matched' in the dataframe. Fill with matched where column 'Pre' = 'pre' and 'Post' = 'post'
merged_df$Matched <- ifelse(merged_df$Pre == 'pre' & merged_df$Post == 'post', 'matched', '')

head(merged_df)


id       Finished.x RecordedDate.x   Q1_a             
1 ._O_04_A TRUE       2022-11-21 17:38 Agree            
2 6_9_02_L TRUE       2023-05-01 12:23 Strongly disagree
3 6_9_12_L TRUE       2023-05-30 8:49  Disagree         
4 6_E_07_A   NA       NA               NA               
5 :_!_09_K TRUE       2023-01-25 9:08  Strongly disagree
6 A_8_12_1   NA       NA               NA               
  Q2_a                       Q3_a              Q4_a              Pre Finished.y
1 Neither agree nor disagree Somewhat agree    Somewhat agree    pre   NA      
2 Strongly disagree          Strongly disagree Strongly disagree pre   NA      
3 Disagree                   Disagree          Disagree          pre   NA      
4 NA                         NA                NA                NA  TRUE      
5 Strongly disagree          Strongly disagree Strongly disagree pre   NA      
6 NA                         NA                NA                NA  TRUE      
  RecordedDate.y   Q1_b              Q2_b              Q3_b             
1 NA               NA                NA                NA               
2 NA               NA                NA                NA               
3 NA               NA                NA                NA               
4 2023-05-02 13:54 Strongly disagree Strongly disagree Strongly disagree
5 NA               NA                NA                NA               
6 2023-05-09 9:49  Strongly disagree Disagree          Strongly disagree
  Q4_b                       Q5                Q6               
1 NA                         NA                NA               
2 NA                         NA                NA               
3 NA                         NA                NA               
4 Neither agree nor disagree Strongly agree    Strongly agree   
5 NA                         NA                NA               
6 Strongly disagree          Strongly disagree Strongly disagree
  Q7                Q8 Post Matched
1 NA                NA NA   NA     
2 NA                NA NA   NA     
3 NA                NA NA   NA     
4 Agree             10 post NA     
5 NA                NA NA   NA     
6 Strongly disagree  0 post NA

In [220]:
library(dplyr)

# Filter rows where 'Matched' column is 'matched'
matched_rows <- merged_df %>%
  filter(Matched == 'matched')

head(matched_rows)
cat("There are ", nrow(matched_rows), "macthing survery responses.")

id       Finished.x RecordedDate.x   Q1_a              Q2_a          
1 A_A_01_A TRUE       2023-05-15 14:05 Agree             Agree         
2 A_A_01_A TRUE       2022-11-21 17:38 Somewhat agree    Disagree      
3 A_A_01_C TRUE       2023-05-01 12:21 Somewhat disagree Disagree      
4 A_A_01_D TRUE       2023-05-02 14:13 Agree             Somewhat agree
5 A_A_01_M TRUE       2023-03-29 20:42 Agree             Strongly agree
6 A_A_01_M TRUE       2023-04-24 16:19 Somewhat disagree Somewhat agree
  Q3_a                       Q4_a              Pre Finished.y RecordedDate.y  
1 Agree                      Agree             pre TRUE       2023-03-07 16:20
2 Somewhat agree             Somewhat agree    pre TRUE       2023-03-07 16:20
3 Somewhat agree             Somewhat disagree pre TRUE       2023-05-01 13:02
4 Somewhat agree             Somewhat disagree pre TRUE       2023-05-02 14:58
5 Somewhat agree             Somewhat disagree pre TRUE       2023-04-24 17:01
6 Neither agree nor disagree Somewhat disagree pre TRUE       2023-04-24 17:01
  Q1_b              Q2_b                       Q3_b                      
1 Strongly disagree Agree                      Somewhat agree            
2 Strongly disagree Agree                      Somewhat agree            
3 Somewhat disagree Neither agree nor disagree Neither agree nor disagree
4 Agree             Somewhat agree             Neither agree nor disagree
5 Somewhat agree    Somewhat agree             Neither agree nor disagree
6 Somewhat agree    Somewhat agree             Neither agree nor disagree
  Q4_b                       Q5                Q6                        
1 Somewhat disagree          Somewhat disagree Somewhat agree            
2 Somewhat disagree          Somewhat disagree Somewhat agree            
3 Neither agree nor disagree Agree             Strongly agree            
4 Neither agree nor disagree Somewhat agree    Neither agree nor disagree
5 Neither agree nor disagree Somewhat agree    Neither agree nor disagree
6 Neither agree nor disagree Somewhat agree    Neither agree nor disagree
  Q7             Q8 Post Matched
1 Agree           9 post matched
2 Agree           9 post matched
3 Strongly agree 10 post matched
4 Somewhat agree  7 post matched
5 Somewhat agree  7 post matched
6 Somewhat agree  7 post matched

There are  3391 macthing survery responses.

In [221]:
final_df <- merged_df %>%
  filter(Matched == 'matched')

cat("There are", nrow(final_df), "survey responses in the final dataset.")

There are 3391 survey responses in the final dataset.

>> 5. Convert likert scale data to numerical format for analysis. 

In [224]:
head(final_df)

id       Finished.x RecordedDate.x   Q1_a              Q2_a          
1 A_A_01_A TRUE       2023-05-15 14:05 Agree             Agree         
2 A_A_01_A TRUE       2022-11-21 17:38 Somewhat agree    Disagree      
3 A_A_01_C TRUE       2023-05-01 12:21 Somewhat disagree Disagree      
4 A_A_01_D TRUE       2023-05-02 14:13 Agree             Somewhat agree
5 A_A_01_M TRUE       2023-03-29 20:42 Agree             Strongly agree
6 A_A_01_M TRUE       2023-04-24 16:19 Somewhat disagree Somewhat agree
  Q3_a                       Q4_a              Pre Finished.y RecordedDate.y  
1 Agree                      Agree             pre TRUE       2023-03-07 16:20
2 Somewhat agree             Somewhat agree    pre TRUE       2023-03-07 16:20
3 Somewhat agree             Somewhat disagree pre TRUE       2023-05-01 13:02
4 Somewhat agree             Somewhat disagree pre TRUE       2023-05-02 14:58
5 Somewhat agree             Somewhat disagree pre TRUE       2023-04-24 17:01
6 Neither agree nor disagree Somewhat disagree pre TRUE       2023-04-24 17:01
  Q1_b              Q2_b                       Q3_b                      
1 Strongly disagree Agree                      Somewhat agree            
2 Strongly disagree Agree                      Somewhat agree            
3 Somewhat disagree Neither agree nor disagree Neither agree nor disagree
4 Agree             Somewhat agree             Neither agree nor disagree
5 Somewhat agree    Somewhat agree             Neither agree nor disagree
6 Somewhat agree    Somewhat agree             Neither agree nor disagree
  Q4_b                       Q5                Q6                        
1 Somewhat disagree          Somewhat disagree Somewhat agree            
2 Somewhat disagree          Somewhat disagree Somewhat agree            
3 Neither agree nor disagree Agree             Strongly agree            
4 Neither agree nor disagree Somewhat agree    Neither agree nor disagree
5 Neither agree nor disagree Somewhat agree    Neither agree nor disagree
6 Neither agree nor disagree Somewhat agree    Neither agree nor disagree
  Q7             Q8 Post Matched
1 Agree           9 post matched
2 Agree           9 post matched
3 Strongly agree 10 post matched
4 Somewhat agree  7 post matched
5 Somewhat agree  7 post matched
6 Somewhat agree  7 post matched

In [227]:
#QC check for completeness of data. Check subset of columns for blank or NA values
blank_or_na_rows <- final_df %>%
  filter(Q1_a == "" | is.na(Q1_a) |
         Q2_a == "" | is.na(Q2_a) |
         Q3_a == "" | is.na(Q3_a) |
         Q4_a == "" | is.na(Q4_a) |
         Q1_b == "" | is.na(Q1_b) |
         Q2_b == "" | is.na(Q2_b) |
         Q3_b == "" | is.na(Q3_b) |
         Q4_b == "" | is.na(Q4_b) |
         Q5 == "" | is.na(Q5) |
         Q6 == "" | is.na(Q6) |
         Q7 == "" | is.na(Q7) |
         Q8 == "" | is.na(Q8))

blank_or_na_rows

id Finished.x RecordedDate.x Q1_a Q2_a Q3_a Q4_a Pre Finished.y
     RecordedDate.y Q1_b Q2_b Q3_b Q4_b Q5 Q6 Q7 Q8 Post Matched

In [228]:
# get unique values of all columns
unique_values <- lapply(final_df, unique)

unique_values


$id
   [1] "A_A_01_A"                                                             
   [2] "A_A_01_C"                                                             
   [3] "A_A_01_D"                                                             
   [4] "A_A_01_M"                                                             
   [5] "A_A_02_A"                                                             
   [6] "A_A_02_C"                                                             
   [7] "A_A_02_I"                                                             
   [8] "A_A_02_S"                                                             
   [9] "A_A_03_C"                                                             
  [10] "A_A_03_G"                                                             
  [11] "A_A_03_M"                                                             
  [12] "A_A_03_S"                                                             
  [13] "A_A_03_W"                                                             
  [14] "A_A_04_B"                                                             
  [15] "A_A_04_F"                                                             
  [16] "A_A_04_J"                                                             
  [17] "A_A_04_L"                                                             
  [18] "A_A_04_M"                                                             
  [19] "A_A_04_N"                                                             
  [20] "A_A_04_P"                                                             
  [21] "A_A_05_B"                                                             
  [22] "A_A_05_S"                                                             
  [23] "A_A_05_T"                                                             
  [24] "A_A_05_Y"                                                             
  [25] "A_A_06_A"                                                             
  [26] "A_A_06_D"                                                             
  [27] "A_A_06_W"                                                             
  [28] "A_A_07_8"                                                             
  [29] "A_A_07_B"                                                             
  [30] "A_A_07_P"                                                             
  [31] "A_A_07_R"                                                             
  [32] "A_A_07_S"                                                             
  [33] "A_A_08_B"                                                             
  [34] "A_A_08_E"                                                             
  [35] "A_A_08_G"                                                             
  [36] "A_A_08_M"                                                             
  [37] "A_A_08_Z"                                                             
  [38] "A_A_09_C"                                                             
  [39] "A_A_09_F"                                                             
  [40] "A_A_10_A"                                                             
  [41] "A_A_10_B"                                                             
  [42] "A_A_10_C"                                                             
  [43] "A_A_10_S"                                                             
  [44] "A_A_10_T"                                                             
  [45] "A_A_11_H"                                                             
  [46] "A_A_11_I"                                                             
  [47] "A_A_11_L"                                                             
  [48] "A_A_11_W"                                                             
  [49] "A_A_12_A"                                                             
  [50] "A_A_12_C"                                                             
  [51] "A_A_12_F"                             

`7 point linkert scale used`
```python
scale = {
    'Strongly disagree': 1,
    'Disagree': 2,
    'Somewhat disagree': 3,
    'Neither agree nor disagree': 4,
    'Somewhat agree': 5,
    'Agree': 6,
    'Strongly agree': 7
}
```


In [231]:
library(dplyr)

# Define a function to map responses to numerical values
map_responses <- function(response) {
  mapping <- c('Strongly disagree' = 1, # nolint
               'Disagree' = 2, # nolint
               'Somewhat disagree' = 3,
               'Neither agree nor disagree' = 4,
               'Somewhat agree' = 5,
               'Agree' = 6,
               'Strongly agree' = 7)
  return(mapping[response])
}

# Apply the mapping function to the specified columns
final_df <- final_df %>%
  mutate(across(c(Q1_a,
                Q2_a,
                Q3_a,
                Q4_a,
                Q1_b,
                Q2_b,
                Q3_b,
                Q4_b,
                Q5,
                Q6,
                Q7),
                map_responses))

In [232]:
head(final_df)

id       Finished.x RecordedDate.x   Q1_a Q2_a Q3_a Q4_a Pre Finished.y
1 A_A_01_A TRUE       2023-05-15 14:05 6    6    6    6    pre TRUE      
2 A_A_01_A TRUE       2022-11-21 17:38 5    2    5    5    pre TRUE      
3 A_A_01_C TRUE       2023-05-01 12:21 3    2    5    3    pre TRUE      
4 A_A_01_D TRUE       2023-05-02 14:13 6    5    5    3    pre TRUE      
5 A_A_01_M TRUE       2023-03-29 20:42 6    7    5    3    pre TRUE      
6 A_A_01_M TRUE       2023-04-24 16:19 3    5    4    3    pre TRUE      
  RecordedDate.y   Q1_b Q2_b Q3_b Q4_b Q5 Q6 Q7 Q8 Post Matched
1 2023-03-07 16:20 1    6    5    3    3  5  6   9 post matched
2 2023-03-07 16:20 1    6    5    3    3  5  6   9 post matched
3 2023-05-01 13:02 3    4    4    4    6  7  7  10 post matched
4 2023-05-02 14:58 6    5    4    4    5  4  5   7 post matched
5 2023-04-24 17:01 5    5    4    4    5  4  5   7 post matched
6 2023-04-24 17:01 5    5    4    4    5  4  5   7 post matched

In [223]:
# write df to csv
write.csv(final_df, "/Users/patrickokwir/Desktop/Lighthouse-data-notes/Jack-Talk-Surveys/data_output/final_df.csv", row.names = FALSE)

In [233]:
# write df to csv
write.csv(final_df, "/Users/patrickokwir/Desktop/Lighthouse-data-notes/Jack-Talk-Surveys/data_output/final_df_numeric.csv", row.names = FALSE)